### PR TITLE
[distributed] Remove eager wait() in all_gather_tensor

### LINKED ISSUE #172214: [distributed] Remove eager `wait()` in `all_gather_tensor` when `gather_dim != 0`
The `all_gather_tensor` and `all_gather_tensor_autograd` functions currently call `wait()` eagerly when `gather_dim != 0`, which defeats the purpose of `AsyncCollectiveTensor`'s lazy synchronization.

Currently in https://github.com/pytorch/pytorch/blob/1c3e695f1986a14a1eeb85b0ff586196441aff2b/torch/distributed/_functional_collectives.py#L206-L213

So here the explicit wait() call forces synchronization immediately, even though:
- view operations don't need to wait (they can stay wrapped) because `AsyncCollectiveTensor.__torch_dispatch__` already handles automatic waiting when operations need actual data

so it makes async performance worse when using `gather_dim != 0`

we can remove the explicit `wait()` call and let `AsyncCollectiveTensor.__torch_dispatch__` handle it automatically.

So it will wait only triggered when `torch.cat()` actually needs the data and view operations stay wrapped (no wait until data accessed). Also, it will be consistent with async semantics elsewhere. 

Please let me know how does it sound I will come up with a PR soon, thanks!

cc: @wconstab @d4l3k @albanD @fduwjj 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### PR DESCRIPTION
**Pull request by RohitRathore1**
when gather_dim != 0

This implements the TODO to delay the wait() call until the tensor is actually used, enabling better async performance.

Previously, all_gather_tensor would explicitly call wait() on the AsyncCollectiveTensor when gather_dim != 0, which defeated the purpose of lazy synchronization. Now the wait is automatically handled by AsyncCollectiveTensor's __torch_dispatch__ when operations like torch.cat actually need the tensor data. For view-only operations, the wrapper is preserved and the wait is delayed even longer.

To make this work, I also added enable_reentrant_dispatch() in AsyncCollectiveTensor.__torch_dispatch__ so that operations go through the dispatch stack properly, and added AsyncCollectiveTensor support to DebugMode so that debugging/tracing tools can see these operations.

Fixes #172214

### CODE REVIEW

**Review by ezyang — DISMISSED** [review]
spurious kineto update

**Review by zpcore — DISMISSED** [review]
LGTM!

**Comment by zpcore** [comment]
@RohitRathore1 , I saw this is the regression error from the test:
```
REGRESSION: benchmark ('dtensor_dispatch_collectives', 'instruction_count') failed, actual result 38718394 is 53.04% higher than expected 25300000
```
, which I think is triggered here https://github.com/pytorch/pytorch/blob/d7f447c8b1cff3aae568d087efc5e0dbccc4f933/benchmarks/dynamo/pr_time_benchmarks/benchmarks/dtensor.py#L73.

I don't know why delaying the wait() call can trigger such regression. Maybe we can compare the profile.

**Comment by RohitRathore1** [comment]
@zpcore sorry for the confusion in my earlier message.. The overhead was from https://github.com/pytorch/pytorch/blob/d7f447c8b1cff3aae568d087efc5e0dbccc4f933/torch/_utils.py#L1194 

These aren't view ops, so they go through `__torch_dispatch__` which has overhead (tree_map_only, wrapper creation). The fix calls `wait()` early only when we know the fallback path will be used, skipping that dispatch overhead.

File: torch/utils/_debug_mode.py [hunk 1]
```diff
@@ -966,6 +967,17 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
                     stack=self.record_stack_trace,
                 )
                 self._record_call(call)
+        elif AsyncCollectiveTensor in types:
+            # Record AsyncCollectiveTensor operations so debugging/tracing tools can see them
+            if self.record_realtensor:
```
Comment (wconstab) [inline torch/utils/_debug_mode.py:972]: can you show an example print of how this looks when you run the debug tool?

Comment (RohitRathore1) [reply torch/utils/_debug_mode.py:972]: view optimization case (async preserved, no explicit wait_tensor):    
```                                                                                                         
_c10d_functional::all_gather_into_tensor(t: f32[1, 8], 4, 0)  ->  t: f32[4, 8]                                                                                                 
_c10d_functional::_wrap_tensor_autograd(t: f32[4, 8])  ->  t: f32[4, 8]                                                                                                        
aten::view(t: f32[4, 8], [1, 32])  ->  t: f32[1, 32]                                                                                                                           
aten::sum(t: f32[1, 32])  ->  t: f32[]
```                                                                                                                                         
                                                                                                                                                                               
fallback case (wait called early before split/cat):   
```                                                                                                                         
_c10d_functional::all_gather_into_tensor(t: f32[2, 8], 4, 0)  ->  t: f32[8, 8]                                                                                                 
_c10d_functional::_wrap_tensor_autograd(t: f32[8, 8])  ->  t: f32[8, 8]                                                                                                        
_c10d_functional::wait_tensor(t: f32[8, 8])  ->  t: f32[8, 8]                                                                                                                  
aten::split.Tensor(t: f32[8, 8], 2)  ->  [...]                                                                                                                                 
aten::cat([...], 1)  ->  t: f32[2, 32]                                                                                                                                         
aten::sum(t: f32[2, 32])  ->  t: f32[]
```

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -203,13 +204,16 @@ def all_gather_tensor(
         self, group_size, group_name
     )
     res = _maybe_wrap_tensor(tensor)
-    # TODO this should be done inside AsyncCollectiveTensor to delay the wait() call
     if gather_dim != 0:
-        # torch.cat access the data so we already need to wait here, first do wait
-        # and then chunk + cat avoid us going through ACT dispatching logic again
+        # Check if _maybe_view_chunk_cat can use the view optimization.
```
Comment (wconstab) [inline torch/distributed/_functional_collectives.py:209]: Can you explain why we had this wait() conditioned on gather dim != 0 in the first place? 

Any why we need to be so careful to use AsyncCollectiveTensor only when profitable? Is this due to performance (reducing overhead when not needed) or something else?

Comment (RohitRathore1) [reply torch/distributed/_functional_collectives.py:209]: when `gather_dim == 0`, the `all_gather` result is already in the correct layout so no rearrangement needed. The `AsyncCollectiveTensor` is returned directly and wait is naturally deferred until the tensor is consumed but when `gather_dim != 0`, the data needs rearranging `via _maybe_view_chunk_cat`, which has two paths: view path i.e., just `res.view()`, no data movement, and fallback path i.e., `torch.cat(torch.chunk(...))`, needs actual data.

but the original code called `wait()` eagerly for `all gather_dim != 0` cases because of the fallback path. It is purely about performance. `__torch_dispatch__` overhead (tree_map_only for arg unwrapping/rewrapping, wrapper allocation) is measurable - the benchmark showed +53% instruction count (came from the ci failure log). We only preserve the `AsyncCollectiveTensor` wrapper when it provides a real async benefit (view path), and call `wait()` early when the data is needed anyway (fallback path)

**Review by wconstab — APPROVED** [review]
looks ok to me.  can you check test coverage? it might be worth adding a test that uses comm counter to detect wait_tensor ops (does it currently capture these?) and validate the optimization is taking effect

**Comment by RohitRathore1** [comment]
@wconstab The 1st failure is unrealted to my PR. The test_serialization_4gb_file test is being OOM-killed (exit code -9) due to insufficient memory on the CI runner. My changes don't touch serialization code at all. Same for the 2nd one. Please can you help me to force merge it. 😇

**Comment by RohitRathore1** [comment]
> those 2 failures sound OK to bypass with -i. But the latest CI shows lint failures, we'll have to fix those. I'll approve and run now in case you already fixed those, otherwise please fix.

I already fixed lint failures but there is some runtime error in the job so it’s seems out of my hands.

**Comment by wconstab** [comment]
you can check the HUD (https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50) to see the status of a particular CI job on other baselines.  I would suggest basing your work off viable/strict branch if you aren't already.  You probably need to rebase again to be able to get a clean CI signal.

**Comment by jeffdaily** [comment]
Seems it's happening on all platforms.  No idea why this wasn't caught earlier.  Perhaps the CI target determination didn't connect this PR change to the test file test/dynamo/test_structured_trace.py.
